activerecord/lib/active_record/model_schema.rb

```diff
@@ -284,7 +284,9 @@ def quoted_table_name
 
       # Computes the table name, (re)sets it internally, and returns it.
       def reset_table_name # :nodoc:
-        self.table_name = if abstract_class?
+        self.table_name = if self == Base
+          nil
+        elsif abstract_class?
           superclass == Base ? nil : superclass.table_name
```

Comment:
Since we've hoisted the base case, we can simplify this:

```suggestion
          superclass.table_name
```